[CBRD-25717] Support PARALLEL HASH JOIN

youngjinj opened the pull request:
http://jira.cubrid.org/browse/CBRD-25717

## Purpose

파티션 해시 조인 수행 시, 파티션 분할과 각 파티션의 해시 조인을 개별 스레드에서 병렬로 수행합니다.

## Implementation

### 1. PARALLEL 힌트 사용

- `NO_PARALLEL_HASH_JOIN` 힌트 추가

### 2. 파티션 해시 조인 최적화 및 병렬 처리

1. Outer/Inner가 분할된 각 파티션 결과의 첫 번째 컬럼에 해시 키 저장

2. 파티션 분할 관련 함수 분리
   - AS-IS
      - `hjoin_make_partition`
      - `hjoin_make_partition_input`
   - TO-BE
      - `hjoin_try_partition`: 파티션 분할을 수행하는 메인 함수
      - `hjoin_check_partition`: 파티션 분할이 필요한지 여부를 검사
      - `hjoin_prepare_partition`: 파티션 분할에 필요한 공간을 준비
      - `hjoin_init_split_info`, `hjoin_clear_split_info`: 파티션 분할에 필요한 정보를 초기화 및 해제
      - `hjoin_init_shared_split_info`, `clear_shared_split_info`: 병렬 파티션 분할에 필요한 정보를 초기화 및 해제
      - `hjoin_build_partitions`: 병렬 처리를 하지 않는 경우, Outer/Inner 결과를 파티션별로 분할
      - `hjoin_split_qlist`: 병렬 처리를 하지 않는 경우, 입력 결과를 파티션별로 분할
      - `hjoin_merge_qlist`: 분할된 파티션의 조인 결과를 하나로 병합

3. 함수 이름 변경 및 제거
   - 이름 변경
      - `hjoin_partition` -> `hjoin_execute_partitions`
      - `hjoin_with_context` -> `hjoin_execute`
      - `hjoin_internal` -> `hjoin_execute_internal`
   - 제거
      - `hjoin_check_valid_part_info`

4. 병렬 스레드 풀 관리를 위한 예약자 추가
   - `class worker_manager_reserver`: `worker_manager_global`에 필요한 병렬 스레드 수를 예약

5. 병렬 파티션 해시 조인 관련 코드를 `src/query/parallel/px_hash_join`에 추가
   - `px_hash_join.hpp, cpp`
      - `class entry_manager`: 병렬 스레드 생성, 재사용, 종료 시 필요한 작업을 수행
      - `class worker_pool_manager`: 동일한 Task Queue를 공유하는 병렬 스레드 풀 관리
      - `build_partitions`: Outer 및 Inner에 대한 `split_task`를 병렬 스레드 수만큼 생성하고 Task Queue에서 수행
      - `execute_partitions`: 파티션에 대한 `join_task`를 병렬 스레드 수만큼 생성하고 Task Queue에서 수행
   - `px_hash_join_task_manager.hpp, cpp`
      - `class task_manager`: 전체 Task의 완료 여부 확인 및 오류 발생 시 처리
      - `class base_task`: Task 수행을 위한 기본 클래스
      - `class split_task`: 파티션 분할 작업을 병렬 스레드에 할당하여 수행
      - `class join_task`: 파티션 해시 조인 작업을 병렬 스레드에 할당하여 수행
   - `px_hash_join_spawn_manager.hpp, cpp`
      - `class spawn_manager`: 'PRED_EXPR', 'REGU_VARIABLE', 'DB_VALUIE'를 스레드 단위로 복제

6. 해시 테이블 생성 시, 2의 제곱 수보다 큰 가장 가까운 소수를 크기로 사용
   - `mht_calculate_htsize_for_pow2` 함수 추가
   - 해시 테이블의 적재율(Load Factor)이 `MHT_REHASH_TRESHOLD`(70%)를 넘지 않도록 제한하여, 충돌률(Collision Rate)을 30% 이하로 유지하기 위함
      - 입력 튜플 수가 `MHT_REHASH_TRESHOLD`(70%)를 넘지 않도록 해시 테이블 크기 계산
      - 계산된 크기보다 큰 2의 제곱 수 선택
      - 해당 2의 제곱 수보다 큰 가장 가까운 소수를 해시 테이블 크기로 사용

7. `QFILE_LIST_ID`를 연결하여 조인 결과 병합 (`qfile_connect_list`)
   - 조인 결과 병합 방법 비교
      - `qfile_combine_two_list`: 레코드 단위로 복사하여 병합
      - `qfile_append_list`: 페이지 단위로 복사하여 병합
      - `qfile_connect_list`: `QFILE_LIST_ID`를 연결하여 병합

8. SQL Trace 출력 변경
   - `PARTITIONING` 출력 이름을 `SPLIT`으로 변경
   - `part_time`, `build_time`, `probe_time`, `fetch_time`, `skew` 항목 제거
   - Outer/Inner 결과 조회에 대한 출력을 `SUBQUERY (uncorrelated)`로 분리
   - `OUTER`, `INNER` 출력 제거
   - `BUILD` 출력에서 `build: inner/outer` 항목 제거
   - `PROBE` 출력에서 `max_collisions: #` 항목 제거
   - 병렬 파티션 해시 조인 수행 시 `BUILD`, `PROBE` 출력을 `PARALLEL` 하위로 출력

```
Trace Statistics:
  SELECT (time: 27788, fetch: 2021533, fetch_time: 1634, ioread: 0, parallel workers: 2)
    SCAN (temp time: 2027, fetch: 123307, ioread: 0, readrows: 50000000, rows: 50000000)
      HASHJOIN (time: 23279, fetch: 1898224, fetch_time: 1601, ioread: 0, parallel workers: 4)
        SPLIT (time: 3904, fetch: 980009, ioread: 0, partitions: 328)
        PARALLEL (time: 11363, fetch: 572958, ioread: 0)
          BUILD (time: 20..45, fetch: 162138, ioread: 0, rows: 50000000, method: memory, collision_rate: 24%)
          PROBE (time: 54..94, fetch: 409508, ioread: 0, readrows: 50000000, readkeys: 50000000, rows: 50000000)
        SUBQUERY (uncorrelated)
          SELECT (time: 7953, fetch: 344537, fetch_time: 447, ioread: 0)
            SCAN (table: dba.t1), (heap time: 7953, fetch: 344536, ioread: 0, readrows: 50000000, rows: 50000000)
                 (parallel workers: 4, heap time: 7952..7952, readrows: 12269088..12660305, rows: 12269088..12660305, gather: mergeable list)
          SELECT (time: 7936, fetch: 344541, fetch_time: 455, ioread: 0)
            SCAN (table: dba.t1), (heap time: 7936, fetch: 344540, ioread: 0, readrows: 50000000, rows: 50000000)
                 (parallel workers: 4, heap time: 7936..7936, readrows: 11332620..15302952, rows: 11332620..15302952, gather: mergeable list)
```

### 3. `qexec_hash_join` 실행흐름

```cpp
01. 'HASHJOIN_MANAGER' 초기화 ('hjoin_init_manager')
02. 비어 있는 입력 결과가 있는지 확인 ('hjoin_check_empty_inputs')
03. HASHJOIN_STATUS 확인
  04. HASHJOIN_STATUS_FILL_NULL_VALUES ('hjoin_outer_fill_null_values')
    - Outer Join의 경우, Preserved-Side의 결과를 유지하고, Null-Supplying을 NULL로 채움
  05. HASHJOIN_STATUS_TRY ('hjoin_try_partition')
    06. 파티션 해시 조인이 필요한지 여부를 검사 ('hjoin_check_partition')
    07. 파티션 해시 조인이 필요한 경우
      08. 파티션 분할 공간 준비 ('hjoin_prepare_partition')
      09. 병렬 파티션 해시 조인이 가능한지 검사 ('hjoin_try_parallel')
      10. HASHJOIN_STATUS 확인
        11. HASHJOIN_STATUS_PARTITION ('hjoin_build_partitions')
          - 싱글 스레드로 파티션 분할 
        12. HASHJOIN_STATUS_PARALLEL ('parallel_query::hash_join::build_partitions')
          13. 병렬 스레드로 파티션 분할
            14. Outer 결과를 한 페이지 단위로 나누어 개별 스레드에서 파티션 분할 수행
              - 미리 확보된 첫 번째 컬럼의 빈 공간에 계산된 해시 키 저장
            15. Inner 결과를 한 페이지 단위로 나누어 개별 스레드에서 파티션 분할 수행
              - 미리 확보된 첫 번째 컬럼의 빈 공간에 계산된 해시 키 저장
      16. 완료 시 Outer/Inner 결과 제거(Destroy) ('hjoin_destroy_qlist')
      17. 에러 발생 시 HASHJOIN_STATUS_SINGLE로 Failback
    18. HASHJOIN_STATUS 확인
      19. HASHJOIN_STATUS_SINGLE ('hjoin_execute')
        - 싱글 스레드로 단일 해시 조인 수행
      20. HASHJOIN_STATUS_PARTITION ('hjoin_execute_partitions')
        - 싱글 스레드로 파티션 해시 조인 수행
      21. HASHJOIN_STATUS_PARALLEL ('parallel_query::hash_join::execute_partitions')
        22. 병렬 스레드로 파티션 해시 조인 수행
          23. 'HASHJOIN_CONTEXT' 단위로 나누어 개별 스레드에서 해시 조인 수행
            24. TLS(Thread-Local Storage) 변수가 NULL인 경우, 'cubxasl::spawner' 객체 생성
            25. 'cubxasl::spawner'를 통해 'PRED_EXPR', 'REGU_VARIABLE', 'DB_VALUIE'를 스레드 단위로 복제
            26. 'BUILD': 첫 번째 컬럼에 저장된 해시 키를 사용해 해시 테이블 생성 ('hjoin_build')
            27. 'PROBE': 첫 번째 컬럼에 저장된 해시 키를 사용해 해시 테이블 탐색 ('hjoin_probe', 'hjoin_outer_probe')
              28. 해시 테이블에서 매칭되는 레코드를 찾은 경우
                29. 'BUILD' 레코드에서 조인 컬럼 값을 'DB_VALUE'로 읽음
                30. 'PROBE' 레코드에서 조인 컬럼 값을 'DB_VALUE'로 읽음
                31. 조인 컬럼 값 비교
                32. Outer Join의 경우, During-Join Predicate 평가
                33. 비교 결과가 일치할 경우 조인 결과로 입력
            34. 조인이 완료된 'HASHJOIN_CONTEXT'의 파티션 결과를 제거(Destroy) ('hjoin_destroy_qlist')
          35. 파티션별 조인 결과 병합 ('hjoin_merge_qlist' -> 'qfile_connect_list')
36. 'xasl->list_id'에 결과 담기 ('qfile_copy_list_id')
37. 'HASHJOIN_MANAGER'를 포함하여 사용된 자원 해제 ('hjoin_clear_manager')
```

### 4. `PRED_EXPR`, `REGU_VARIABLE`, `DB_VALUE` 동시성 관리

1. 관련 코드를 `src/xasl/xasl_spawner.hpp` 및 `xasl_spawner.cpp`에 추가
   - `parallel_heap_scan::memory_mapper` 코드를 참고하여 구현
   - Original `XASL_NODE`의 주소를 키로 하여 `std::unordered_map` 기반으로 복제
   - 해시 조인에 사용되지 않는 일부 `XASL_NODE`에 포함된 타입은 복제를 지원하지 않음
      - 해시 조인에서는 입력 결과에 대한 리스트 스캔은 수행되지만, 테이블 스캔은 수행되지 않음

2. 복제본 생성 및 해제
   - 복제본은 병렬 스레드 수만큼 생성된 `join_task`에 의해 생성 및 해제
   - 복제본은 병렬 스레드의 TLS(Thread-Local Storage) 변수로 관리

### 5. 병렬 스레드에 대한 Resource Tracker 동작

1. 병렬 스레드 재사용 시에는 확인하지 않음
   - `cubthread::entry`에 `bool skip_end_resource_tracks_in_recycle` 추가
   - `skip_end_resource_tracks_in_recycle`가 true인 경우 병렬 스레드 종료 시에만 확인
   - `join_task`를 병렬 스레드 수만큼 생성하는 것으로 변경되면서 사용되지 않음

2. 병렬 스레드에서의 `m_qlist_count`의 증감은 메인 스레드에 반영
   - `cubthread::entry`의 `m_px_orig_thread_entry`를 메인 스레드로 설정

### 6. 제약사항 추가

1. Click Couter 함수 사용 시 해시 조인을 수행할 수 없음
   - INCR/DECR
   - WITH INCREMENT/INCREMENT FOR

## Remarks

N/A

---

### SQL Hint
- Added `NO_PARALLEL_HASH_JOIN` hint.

### Hash Table
- Added `mht_calculate_htsize_for_pow2`.
  - Determines hash table size as the nearest prime greater than a power of two.

### Partitioning
1. Store the hash key in the first column of each partitioned outer/inner result.
2. Refactored functions:
   - `hjoin_try_partition`
   - `hjoin_check_partition`
   - `hjoin_prepare_partition`
   - `hjoin_init_split_info`, `hjoin_clear_split_info`
   - `hjoin_init_shared_split_info`, `clear_shared_split_info`
   - `hjoin_build_partitions`
   - `hjoin_split_qlist`
   - `hjoin_merge_qlist`

### Parallel Execution
1. Added `worker_manager_reserver` for reserving parallel workers in `worker_manager_global`.
   - `class worker_manager_reserver`
2. New directory `src/query/parallel/px_hash_join`:
   - `px_hash_join.hpp, cpp`
      - `class entry_manager`
      - `class worker_pool_manager`
      - `build_partitions`
      - `execute_partitions`
   - `px_hash_join_task_manager.hpp, cpp`
      - `class task_manager`
      - `class base_task`
      - `class split_task`
      - `class join_task`
   - `px_hash_join_spawn_manager.hpp, cpp`
      - `class spawn_manager`
3. Concurrency Management for `PRED_EXPR`, `REGU_VARIABLE`, `DB_VALUE`
   - Implemented in `src/xasl/xasl_spawner.hpp` and `xasl_spawner.cpp`.
   - Each thread creates its own clone for independent use.
   - Clones are allocated and freed by `join_task` and stored in thread-local storage (TLS).

### Resource Tracker
1. Skip resource tracking on parallel thread recycle.
   - Added `bool skip_end_resource_tracks_in_recycle` to `cubthread::entry`.
   - Checked only on thread termination when the flag is set.
2. Updates to `m_qlist_count` in parallel threads are reflected in the main thread.
   - Set `m_px_orig_thread_entry` in `cubthread::entry` to the main thread.

### Merging Join Results from Partitions
- Merge join results by connecting `QFILE_LIST_ID` (`qfile_connect_list`) instead of copying records or pages.

### SQL Trace
1. Renamed `PARTITIONING` output to `SPLIT`.
2. Removed `part_time`, `build_time`, `probe_time`, `fetch_time`, `skew`.
3. Separated outer/inner result scans into `SUBQUERY (uncorrelated)`.
4. Removed `OUTER` and `INNER` outputs.
5. In `BUILD` output, removed `build: inner/outer`.
6. In `PROBE` output, removed `max_collisions: #`.
7. For parallel partitioned hash join, moved `BUILD` and `PROBE` under `PARALLEL`.

### Restrictions
- Hash join is not allowed when using Click Counter functions:
   - INCR/DECR
   - WITH INCREMENT/INCREMENT FOR

youngjinj committed:
prototype 구현

youngjinj committed:
parallel_query::worker_manager_with_dedicated_pool 사용

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into HEAD

youngjinj committed:
parallel_query::worker_manager_with_dedicated_pool::get_manager().join(); 추가 (임시)

youngjinj committed:
Parallel Hash Join에 대한 SQL Trace 출력 문제 해결

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
디버그 매크로 비활성화

youngjinj committed:
SQL Trace 용어 변경

youngjinj committed:
인터럽트 및 에러 발생 시 처리 코드 추가

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
px_hash_join.hpp, cpp 분리

youngjinj committed:
px_hash_join.hpp, cpp 에 세부적인 구현 추가

youngjinj committed:
parallel_hash_join 에서 사용을 위한 static 함수 변경

youngjinj committed:
불필요한 변경사항 제거

youngjinj committed:
xasl_spawner.hpp, cpp 추가 (XASL 부분 복제)

youngjinj committed:
xasl_spawner.hpp, cpp 에서 예외처리 코드 정리

youngjinj committed:
cubxasl::spawner 를 Parallel Thread 단위로 생성

youngjinj committed:
custom_entey_manager 추가 (cubthread::entry_manager 변경 제거)

youngjinj committed:
parallel thread는 recycle_context에서 end_resource_tracks를 하지 않음

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
잘못된 변경사항 수정

youngjinj committed:
click count 함수 사용 시 해시 조인 실행계획 포기

youngjinj committed:
parallel_query::hash_join::entry_manager 내부에서 emulate_tid 초기화

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
px_hash_join.hpp, cpp 디렉터리 이동

youngjinj committed:
누락된 변경사항

youngjinj committed:
worker_manager_reserver, worker_pool_manager 추가

youngjinj committed:
HASHJOIN_FETCH_INFO 관련 정보 통합, PT_HINT_PARALLEL 힌트 사용

youngjinj committed:
코드 주석 변경

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
cubthread::entry 에 skip_end_resource_tracks_in_recycle 추가

youngjinj committed:
해시 테이블 충돌 비율 관련 변경

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
Parallel 스레드의 SQL Trace 정보가 Main 스레드에서 누락되는 문제 해결

youngjinj committed:
split 병렬처리

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
pgbuf_fix 최소화. dependent_list_id 추가

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
버그 수정

youngjinj committed:
split 병렬 처리에 따른 sql trace 코드 정리

youngjinj committed:
shell 테스트를 위한 강제 활성화

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj commented:
/run all

youngjinj commented:
/run shell

youngjinj committed:
HASHJOIN_PROFILE_TIME 설정 시에만 파티션 결과 머지에 대한 Trace 출력

youngjinj committed:
hjoin_outer_probe 에서 누락된 변경사항

youngjinj committed:
SQL Trace 수집 관련 변경사항

youngjinj committed:
NO_PARALLEL_HASH_JOIN 힌트 추가

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
MERGE 문에서 해시 조인의 병렬 처리를 지원하지 않음. statdump 항목 추가.

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj commented:
/run shell

youngjinj committed:
sql 테스트 실패 수정

youngjinj commented:
/run shell

youngjinj committed:
사용하지 않는 변수, 오타 수정

xmilex-git commented on src/thread/thread_entry.cpp:
```diff
@@ -137,6 +137,7 @@ namespace cubthread
     , m_px_lock ()
     , m_px_stats (NULL)
     , m_px_orig_thread_entry (NULL)
+    , skip_end_resource_tracks_in_recycle (false)
```
```suggestion
    , m_skip_end_resource_tracks_in_recycle (false)
```

youngjinj replied:
반영했습니다. (45bf43a)

youngjinj committed:
m_  접두어 추가, 해시 조인 강제 활성화 제거

shparkcubrid commented:
병렬을 16으로 느리면 오히려 느려지는 현상이 있습니다. 확인해주세요.
```
csql> select count (*) from (                                                                                                                                                                           select /*+ recompile use_hash parallel(8) */                                                                                                                                                            a.c1 as a_c1, b.c1 as b_c1                                                                                                                                                                          from t1 a, t1 b                                                                                                                                                                                       where a.c1 = b.c1                                                                                                                                                                                   );

=== <Result of SELECT Command in Line 1> ===

              count(*)
======================
               1000000

1 row selected. (0.264000 sec) Committed. (0.000000 sec)

1 command(s) successfully processed.
csql> select count (*) from (                                                                                                                                                                           select /*+ recompile use_hash parallel(16) */                                                                                                                                                           a.c1 as a_c1, b.c1 as b_c1                                                                                                                                                                          from t1 a, t1 b                                                                                                                                                                                       where a.c1 = b.c1                                                                                                                                                                                   );

=== <Result of SELECT Command in Line 1> ===

              count(*)
======================
               1000000

1 row selected. (0.534000 sec) Committed. (0.000000 sec)
```
그리고 테스트 케이스 등록해주세요.

youngjinj commented:
> 병렬을 16으로 느리면 오히려 느려지는 현상이 있습니다. 확인해주세요.
> 
> ```
> csql> select count (*) from (                                                                                                                                                                           select /*+ recompile use_hash parallel(8) */                                                                                                                                                            a.c1 as a_c1, b.c1 as b_c1                                                                                                                                                                          from t1 a, t1 b                                                                                                                                                                                       where a.c1 = b.c1                                                                                                                                                                                   );
> 
> === <Result of SELECT Command in Line 1> ===
> 
>               count(*)
> ======================
>                1000000
> 
> 1 row selected. (0.264000 sec) Committed. (0.000000 sec)
> 
> 1 command(s) successfully processed.
> csql> select count (*) from (                                                                                                                                                                           select /*+ recompile use_hash parallel(16) */                                                                                                                                                           a.c1 as a_c1, b.c1 as b_c1                                                                                                                                                                          from t1 a, t1 b                                                                                                                                                                                       where a.c1 = b.c1                                                                                                                                                                                   );
> 
> === <Result of SELECT Command in Line 1> ===
> 
>               count(*)
> ======================
>                1000000
> 
> 1 row selected. (0.534000 sec) Committed. (0.000000 sec)
> ```
> 
> 그리고 테스트 케이스 등록해주세요.

먼저, 성능 저하로 느껴지시는 현상에 대해 설명드리겠습니다.

max_parallel_workers 파라미터의 기본값은 32입니다.

`/*+ parallel(8) */` 힌트를 사용할 경우:
Parallel Subquery: 1개 + Parallel Heap Scan: 8 * 2 = 총 17개의 병렬 스레드가 사용됩니다.
이는 최대값 32보다 작기 때문에, 모든 연산이 병렬로 수행됩니다.

반면 `/*+ parallel(16) */` 힌트를 사용할 경우:
Parallel Subquery: 1개 + Parallel Heap Scan: 16 * 2 = 총 33개의 병렬 스레드가 필요합니다.
이는 최대값 32를 초과하므로, 모든 연산을 병렬로 실행할 수 없습니다.
따라서 Outer 또는 Inner 입력 중 한쪽만 병렬 Heap Scan을 수행하게 됩니다.

```sql
drop table if exists t1;
create table t1 (c1 int);
insert into t1
  select * from (
    with recursive cte (n) as (select 1 union all select n + 1 from cte where n < 2000)
    select rownum from cte a, cte b, cte c limit 1000000
  ) order by random ();

set trace on;
```

```sql
select count (*) from (
    select /*+ recompile use_hash parallel(8) */
      a.c1 as a_c1, b.c1 as b_c1
    from t1 a, t1 b
    where a.c1 = b.c1
  );

show trace;
```

```diff
  Trace Statistics:
    SELECT (time: 356, fetch: 47232, fetch_time: 25, ioread: 0, parallel workers: 2)
      SCAN (temp time: 39, fetch: 2467, ioread: 0, readrows: 1000000, rows: 1000000)
        HASHJOIN (time: 254, fetch: 44763, fetch_time: 24, ioread: 0, parallel: 8)
          SPLIT (time: 20, fetch: 19611, ioread: 0, partitions: 7)
          PARALLEL (time: 128, fetch: 11470, ioread: 0)
            BUILD (time: 24..29, fetch: 3244, ioread: 0, rows: 1000000, method: memory, collision_rate: 23%)
            PROBE (time: 52..53, fetch: 8198, ioread: 0, readrows: 1000000, readkeys: 1000000, rows: 1000000)
          SUBQUERY (uncorrelated)
            SELECT (time: 94, fetch: 6826, fetch_time: 7, ioread: 0)
              SCAN (table: dba.t1), (heap time: 94, fetch: 6825, ioread: 0, readrows: 1000000, rows: 1000000)
+                  (parallel workers: 8, heap time: 93..93, readrows: 124093..125802, rows: 124093..125802, gather: mergeable list)
            SELECT (time: 95, fetch: 13666, fetch_time: 15, ioread: 0)
              SCAN (table: dba.t1), (heap time: 95, fetch: 13665, ioread: 0, readrows: 1000000, rows: 1000000)
                   (parallel workers: 8, heap time: 94..94, readrows: 123714..126315, rows: 123714..126315, gather: mergeable list)
```
```sql
select count (*) from (
    select /*+ recompile use_hash parallel(16) */
      a.c1 as a_c1, b.c1 as b_c1
    from t1 a, t1 b
    where a.c1 = b.c1
  );

show trace;
```

```diff
  Trace Statistics:
    SELECT (time: 594, fetch: 47206, fetch_time: 28, ioread: 0, parallel workers: 2)
      SCAN (temp time: 48, fetch: 2467, ioread: 0, readrows: 1000000, rows: 1000000)
        HASHJOIN (time: 495, fetch: 44737, fetch_time: 28, ioread: 0, parallel: 16)
          SPLIT (time: 14, fetch: 19648, ioread: 0, partitions: 7)
          PARALLEL (time: 132, fetch: 11475, ioread: 0)
            BUILD (time: 27..33, fetch: 3237, ioread: 0, rows: 1000000, method: memory, collision_rate: 23%)
            PROBE (time: 53..55, fetch: 8210, ioread: 0, readrows: 1000000, readkeys: 1000000, rows: 1000000)
          SUBQUERY (uncorrelated)
            SELECT (time: 337, fetch: 6892, fetch_time: 1, ioread: 0)
              SCAN (table: dba.t1), (heap time: 162, fetch: 1916, ioread: 0, readrows: 1000000, rows: 1000000)
-                                                                                                             
            SELECT (time: 77, fetch: 6704, fetch_time: 13, ioread: 0)
              SCAN (table: dba.t1), (heap time: 77, fetch: 6703, ioread: 0, readrows: 1000000, rows: 1000000)
                   (parallel workers: 16, heap time: 77..77, readrows: 51678..66294, rows: 51678..66294, gather: mergeable list)
```

테스트 케이스는 이슈에 한 번에 첨부하기 위해서 아직 준비 중입니다.
곧 올리겠습니다.

youngjinj committed:
테스트 케이스 추가 중 발견한 문제 해결

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

xmilex-git commented on src/thread/thread_entry_task.cpp:
```diff
@@ -85,10 +87,28 @@ namespace cubthread
+	int ioreads_offset = pstat_Metadata[PSTAT_PB_NUM_IOREADS].start_offset;
+	int fetch_time_offset = pstat_Metadata[PSTAT_PB_PAGE_FIX_ACQUIRE_TIME_10USEC].start_offset;
+
+	pthread_mutex_lock (&parent_context->m_px_lock);
+	perfmon_add_stat (parent_context, PSTAT_PB_NUM_FETCHES, context.m_px_stats[fetches_offset]);
+	perfmon_add_stat (parent_context, PSTAT_PB_NUM_IOREADS, context.m_px_stats[ioreads_offset]);
+	perfmon_add_at_offset_to_local (parent_context, fetch_time_offset, context.m_px_stats[fetch_time_offset]);
+	pthread_mutex_unlock (&parent_context->m_px_lock);
```
parent 쪽 thread에서 `perfmon_add_stat`을 수행할 때 lock을 잡는 코드가 없는 것 같습니다. 확인 부탁드립니다.

youngjinj replied:
main과 parallel 스레드 사이에 perfmon_add_stat 동시성을 제어할 수 있도록 보완하겠습니다.
예) parallel 스레드는 main 스레드에 바로 perfmon_add_stat 하지 않고, main 스레드의 m_px_stats에 먼저 누적한 후, main 스레드에서 자신의 m_px_stats를 반영

youngjinj replied:
반영했습니다. (48f549f, 7fcc96d)

xmilex-git commented on src/xasl/xasl_spawner.cpp:
```diff
@@ -0,0 +1,933 @@
+    return dest;
+  }
+
+  PL_SIGNATURE_TYPE *
+  spawner::spawn (const PL_SIGNATURE_TYPE *src)
+  {
+    /* TODO: unsupported */
+    assert_release (src == nullptr);
+    return nullptr;
+  }
```
현재 SP가 during join term 등으로 존재하면 parallel hash join의 실행을 parser 수준에서 불가능하게 block 하는 코드가 있나요?

youngjinj replied:
SP에 대한 동시성 처리가 지원되지 않는 것으로 보이기 때문에 실행계획 생성 단계에서 Parallel 실행을 막는 코드를 추가하려고 합니다.

youngjinj replied:
반영했습니다. (7fcc96d)

xmilex-git commented on src/query/parallel/px_hash_join.cpp:
```diff
@@ -0,0 +1,985 @@
+	      tuple_index++;
+
+	      /* overflow page */
+	      if (QFILE_GET_OVERFLOW_PAGE_ID (page) != NULL_PAGEID)
```
```suggestion
	      if (__builtin_expect(QFILE_GET_OVERFLOW_PAGE_ID (page) != NULL_PAGEID, 0))
```
split 시 row-by-row로 overflow page가 있는지 확인한다면 성능 저하가 생길까 염려됩니다. 이 부분에서 성능 부하가 적나요? 또한, `get_next_page (thread_ref)` 이후에 overflow page가 존재하는 페이지라면 아예 다른 루틴을 통해 iteration 하는 것도 최적화할 수 있는 방법이 될 것 같습니다.

youngjinj replied:
따로 의견을 주신대로 오버플로우 페이지에 대한 분기를 do { ... } while 문 밖으로 분리하는 것으로 해결했습니다.
(d45156f)

xmilex-git commented on src/query/parallel/px_hash_join.cpp:
```diff
@@ -0,0 +1,985 @@
+		  break;
+		}
+
+	      if (temp_part_list_id[part_id]->tfile_vfid->membuf_last == temp_part_list_id[part_id]->tfile_vfid->membuf_npages - 1)
```
`membuf_npages = 4` 라면, membuf의 마지막 페이지에 첫번째 row를 `qfile_add_tuple_to_list()` 한 이후 이 루틴을 수행하게 되지 않나요? 이러면 membuf의 마지막 페이지를 거의 활용하지 못할 것 같습니다.

youngjinj replied:
`temp_part_list_id`를 사용하는 목적은 `part_list_id`의 페이지에 대해 `fix`, `unfix` 발생 횟수를 줄이는 데 있습니다.
`temp_part_list_id`를 사용하는 경우에도 메모리 버퍼 페이지만 사용하고, `file_create_temp`는 호출되지 않아야 합니다.

걱정되는 부분은 메모리 버퍼의 마지막 페이지까지 효율적으로 사용하려는 과정에서 `file_create_temp` 호출이 발생하는 것입니다. 이것을 방지하기 위해서 메모리 버퍼의 마지막 페이지를 쓰기 시작할 때 `temp_part_list_id[part_id]`를 `part_list_id[part_id]`로 병합합니다.

xmilex-git replied:
```c++
	      if ((temp_part_list_id[part_id]->tfile_vfid->membuf_last == temp_part_list_id[part_id]->tfile_vfid->membuf_npages - 1)
		  && (tuple_length + temp_part_list_id[part_id]->last_offset) > DB_PAGESIZE)
		{
		  qfile_close_list  (&thread_ref, temp_part_list_id[part_id]);

		  {
		    std::unique_lock lock (m_shared_info->part_mutexes[part_id]);
		    if (part_list_id[part_id]->tuple_cnt > 0)
		      {
			qfile_append_list (&thread_ref, part_list_id[part_id], temp_part_list_id[part_id]);
			qfile_destroy_list (&thread_ref, temp_part_list_id[part_id]);
		      }
		    else
		      {
			qfile_destroy_list (&thread_ref, part_list_id[part_id]);
			qfile_copy_list_id (part_list_id[part_id], temp_part_list_id[part_id], false);
		      }
		  }

		  QFILE_FREE_AND_INIT_LIST_ID (temp_part_list_id[part_id]);
		  temp_part_list_id[part_id] =
			  qfile_open_list (&thread_ref, &list_id->type_list, nullptr, list_id->query_id, QFILE_FLAG_ALL, nullptr);
		  if (temp_part_list_id[part_id] == nullptr)
		    {
		      assert_release (er_errid () != NO_ERROR);
		      m_task_manager.handle_error (thread_ref);
		      has_error = true;
		      break;
		    }
		}

	      error = qfile_add_tuple_to_list (&thread_ref, temp_part_list_id[part_id], tuple_record.tpl);
```
이렇게 수정하면, membuf를 끝까지 다 쓸 수 있습니다. `3개의 page가 찰 때 마다 flush` -> `4개의 page가 찰 때 마다 flush` 를 고려할 만큼 이 쪽 부분에 부하가 있다면 코드를 이렇게 수정할 수도 있을 것 같습니다. 검토만 해주시면 될 것 같아 resolve 처리하겠습니다.

youngjinj replied:
반영했습니다. (d45156f)

xmilex-git commented on src/query/parallel/px_hash_join.cpp:
```diff
@@ -0,0 +1,985 @@
+    thread_local VAL_DESCR *tls_val_descr = nullptr;
+    thread_local PRED_EXPR *tls_during_join_pred = nullptr;
+    thread_local REGU_VARIABLE_LIST tls_outer_regu_list_pred = nullptr;
+    thread_local REGU_VARIABLE_LIST tls_inner_regu_list_pred = nullptr;
```
thread_local 정보들에 대해서 생성자/소멸자 및 접근/할당/해제가 명확하게 관리되게끔 class로 관리되었으면 좋겠습니다. 예를 들면 다음과 같이 작성하면 어떨까요?
```suggestion
    class thread_local_storage
    {
      std::unique_ptr<cubxasl::spawner> spawner;
      VAL_DESCR *val_descr;
      PRED_EXPR *during_join_pred;
      REGU_VARIABLE_LIST outer_regu_list_pred;
      REGU_VARIABLE_LIST inner_regu_list_pred;
    };
    thread_local thread_local_storage *tls_storage;
```

youngjinj replied:
좀 더 고민한 후에 변경 여부를 말씀드리겠습니다.

youngjinj replied:
반영했습니다. ([6f3e74a](https://github.com/CUBRID/cubrid/pull/6247/commits/6f3e74a60ec3a996e6b74b8aead991b03d5587cf))

xmilex-git reviewed (approved):
차후에, `worker_manager_reserver`와 `entry_manager`를 사용하는 구현 대신 `entry_manager`를 지정하여 `worker_manager_with_dedicated_pool`을 사용할 수 있게 코드를 변경하여 코드를 통합하면 좋을 것 같습니다. 수고하셨습니다!

youngjinj committed:
assert_release -> assert_release_error, QFILE_DEPENDENT_MODE 추가

youngjinj committed:
code style

youngjinj committed:
memory_wrapper 문제 해결

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
m_px_stats 를 통한 parallel stat 누적 과정 정리

youngjinj committed:
불필요한 주석 제거

youngjinj committed:
Java SP, PL/CSQL 사용 시 병렬 처리 비활성화. During Join 조건 관련 버그 수정

youngjinj committed:
TEST_HASH_JOIN_FORCE_ENABLE

youngjinj committed:
QO_PLAN_PARALLEL_OPT_USE 관련 버그 수정

youngjinj commented:
/run shell

youngjinj committed:
버그 수정

youngjinj committed:
버그 수정

youngjinj commented:
/run shell

xmilex-git commented on src/query/parallel/px_hash_join.cpp:
```diff
@@ -0,0 +1,1008 @@
+
+      m_worker_pool = cubthread::get_manager()->create_worker_pool (pool_size, pool_size /* meaningless */,
+		      "parallel hash join workers",
+		      &m_entry_manager, 1, false);
```
wait_for_task_time = 0 으로 설정하지 않아도 되나요? retire_context를 기다리는 루틴은 없나요?

youngjinj committed:
TLS 변수 관련 spawn_manager 추가

youngjinj committed:
qfile_connect_list 하기 전에 append_list_id가 이미 포함되어 있는지 검사

sohee-dgist commented on src/query/parallel/px_hash_join.cpp:
```diff
@@ -0,0 +1,1088 @@
+		  memcpy (overflow_record.tpl + copy_offset, (char *) overflow_page + QFILE_PAGE_HEADER_SIZE, copy_size);
+
+		  copy_offset += copy_size;
+		  assert (copy_offset < tuple_length);
```
assert (copy_offset =< tuple_length);
로 바꿔야 할 수도 있어 보입니다.
혹시 같을 경우는 없을까요?

youngjinj replied:
복사를 하고 나서 `copy_offset += copy_size;` 코드를 실행하기 때문에
마지막 페이지라면 같아지는 것으로 보입니다.
좀 더 확인해본 후에 `assert (copy_offset =< tuple_length);`로 변경하겠습니다.

youngjinj replied:
반영했습니다. ([acdfa25](https://github.com/CUBRID/cubrid/pull/6247/commits/acdfa2576f7d492f0f22977b82968a501ebdcf50))

youngjinj committed:
오버플로우 레코드 사이의 조인 시 버그 수정. notify_stop, stop_execution 처리 추가

youngjinj commented:
/run shell

youngjinj committed:
src/query/parallel/px_hash_join 파일 분할, perfmon_merge_parallel_stats 추가

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
perf_monitor 관련 변경사항

youngjinj commented:
/run shell

HyunukLee commented on src/query/query_hash_join.c:
```diff
@@ -945,6 +840,12 @@ hjoin_init_manager (THREAD_ENTRY * thread_p, HASHJOIN_MANAGER * manager, XASL_NO
 	}
     }
 
+  manager->qlist_merge_method = HASHJOIN_MERGE_CONNECT;
+  manager->qlist_flag =
+    (manager->qlist_merge_method == HASHJOIN_MERGE_CONNECT) ? QFILE_FLAG_ALL | QFILE_NOT_USE_MEMBUF : QFILE_FLAG_ALL;
```
코드 상으로는 무조건 QFILE_NOT_USE_MEMBUF를 추가하게 되어 있는 것으로 보이는데 혹시 다른 의도가 있으신 걸까요?

youngjinj replied:
`qlist_merge_method`를 `HASHJOIN_MERGE_CONNECT`로 설정하는 것은 여러 개의 파티션에서 조인된 결과를(`QFILE_LIST_ID`) 하나로 병합할 때, 하나의 `QFILE_LIST_ID` 구조에 다시 쓰는게(copy) 아니라 연결시켜서 병합하겠다는 것입니다. 그러면 서로 다른 `QFILE_LIST_ID`에 있는 페이지들이 페이지 헤더를 통해 연결되는 구조가 됩니다. 이 방법은 페이지 복사 비용을 줄이는 장점이 있습니다.

페이지는 `VPID`를(볼륨ID, 페이지ID) 사용해서 찾습니다. 메모리 버퍼 페이지는 `QFILE_LIST_ID`마다 파라미터에서 정하는 개수를 가지고 있으며, 항상 메모리에 상주하고 있어서 데이터 버퍼를 경유하지 않고 사용할 수 있습니다. 메모리 버퍼를 경유하지 않기 때문에 유일한 `VPID`를 가지고 있지 않아도 되므로 볼륨ID는 -1, 페이지ID는 0부터 순번으로 사용합니다. `QFILE_LIST_ID`가 다르더라도 메모리 버퍼의 `VPID`는 `(-1. 0~#)`로 동일합니다.

여러 개의 `QFILE_LIST_ID`가 연결되어 있을 때, `VPID`가 `(-1, 3)`인 페이지를 찾으려고 한다면, 여러 개의 `QFILE_LIST_ID` 중에 어느 `QFILE_LIST_ID`의 4번째 메모리 버퍼 페이지인지 구분할 수 없습니다. 그래서 `QFILE_LIST_ID`가 연결되는 구조에서는 메모리 버퍼 페이지를 사용하지 않아야 합니다.

`QFILE_LIST_ID`를 만들 때 `QFILE_NOT_USE_MEMBUF` 플래그를 사용하는 것은 최종 조인 결과를 만들 때 메모리 버퍼 페이지를 사용하는 `QFILE_LIST_ID`를 사용하겠다는 것입니다.

HyunukLee replied:
자세히 설명해주셔서 감사합니다. 다만 여쭤보고 싶었던 것은 HASHJOIN_MERGE_CONNECT로 method를 설정해주고 이어서 바로 삼항연산자가 있어서 논리적으로 항상 참이 되는 것 같아 삼항연산자를 사용하신 다른 의도가 있으신건지 궁금합니다.

youngjinj replied:
하드 코딩이 되어 있기는 하지만 디버깅 중에 `qlist_merge_method`를 변경하면서 테스트해볼 수 있는 방법을 마련한 것입니다. Parallel Hash Join이 머지된 후에 `HASHJOIN_MERGE_CONNECT` 방식을 사용하는 것에서 문제가 발생하지 않으면 해당 코드를 제거하는 것을 검토해보겠습니다. 코드가 머지된 후에 `HASHJOIN_MERGE_CONNECT` 방식을 사용하지 못하게 되는 상황이 발생할 수도 있어서 남겨둔 코드로 생각해주시면 감사하겠습니다.

youngjinj replied:
Resolve 하겠습니다.

youngjinj committed:
SQL Trace 누적 오류 수정

youngjinj committed:
SQL Trace 누적 오류 수정 보완

youngjinj commented:
/run shell

youngjinj committed:
SQL Trace 누적 오류 수정 보완2

youngjinj commented:
/run shell

youngjinj committed:
SQL Trace 누적 오류 수정 보완3

youngjinj commented:
/run shell

youngjinj commented:
/run shell

Hamkua commented on src/optimizer/plan_generation.c:
```diff
@@ -2617,6 +2639,7 @@ gen_hashjoin (QO_ENV * env, QO_PLAN * plan, BITSET * pred_set, BITSET * subqueri
     }
+  outer_xasl->parallelism = parallelism;
 
   outer_xasl = gen_outer (env, outer_plan, &EMPTY_SET, NULL, NULL, outer_xasl);
```
debug 에서 아래 쿼리처럼 잘못된 조건절이 있으면 assertion 이 발생합니다.

```sql
with cte1 as ( 
    select /*+ recompile materialize use_hash parallel(2) */ a.c1 as a_c1, b.c1, max(a.c2) as a_c3 
    from t1 a, t1 b 
    where a.c1 = b.c1 and max(a.c2) = b.c2 
    group by a_c1) 
select /*+ recompile */ * from cte1 limit 100;
```

hash join 이 추가된 시점부터 발생하네요.

youngjinj replied:
`gen_hashjoin`에서 에러 발생 여부를 확인할 때, 파서에서 발생한 에러를 확인하지 않아서 발생하는 문제입니다.
파서에서 발생한 에러까지 확인하도록 수정하겠습니다. (e89d9e9)
감사합니다.

youngjinj committed:
해시 조인 강제 활성화 해제

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj committed:
gen_hashjoin에 파서 에러 확인 추가

youngjinj commented:
/run sql medium shell

youngjinj committed:
single hash join에서 build input을 구분하기 위한 SQL Trace 출력 변경

youngjinj commented:
/run sql medium shell

youngjinj committed:
셀프 테스트 중 발생한 에러 수정

youngjinj commented:
/run sql medium shell

youngjinj committed:
파티션 테이블 관련 SQL Trace 출력 오류 수정

youngjinj commented:
/run sql medium shell

youngjinj commented:
/run sql medium shell

youngjinj committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25717

youngjinj commented:
/run sql medium

tw-kang commented:
/run shell
